src/category_theory/monoidal/functor.lean

```diff
@@ -0,0 +1,183 @@
+import category_theory.monoidal.category
+
+open category_theory
+open tactic
+
+universes v₁ v₂ v₃ u₁ u₂ u₃
+
+open category_theory.category
+open category_theory.functor
+
+namespace category_theory
+
+section
+
+open monoidal_category
+
+variables (C : Sort u₁) [𝒞 : monoidal_category.{v₁} C]
+          (D : Sort u₂) [𝒟 : monoidal_category.{v₂} D]
+include 𝒞 𝒟
+
+structure lax_monoidal_functor extends C ⥤ D :=
+-- unit morphism
+(ε               : tensor_unit D ⟶ obj (tensor_unit C))
+-- tensorator
+(μ                : Π X Y : C, (obj X) ⊗ (obj Y) ⟶ obj (X ⊗ Y))
+(μ_natural'       : ∀ (X Y X' Y' : C)
+  (f : X ⟶ Y) (g : X' ⟶ Y'),
+  ((map f) ⊗ (map g)) ≫ μ Y Y' = μ X X' ≫ map (f ⊗ g)
+  . obviously)
+-- associativity of the tensorator
+(associativity'   : ∀ (X Y Z : C),
+    (μ X Y ⊗ 𝟙 (obj Z)) ≫ μ (X ⊗ Y) Z ≫ map (associator X Y Z).hom
+  = (associator (obj X) (obj Y) (obj Z)).hom ≫ (𝟙 (obj X) ⊗ μ Y Z) ≫ μ X (Y ⊗ Z)
+  . obviously)
+-- unitality
+(left_unitality'  : ∀ X : C,
+    (left_unitor (obj X)).hom
+  = (ε ⊗ 𝟙 (obj X)) ≫ μ (tensor_unit C) X ≫ map (left_unitor X).hom
+  . obviously)
+(right_unitality' : ∀ X : C,
+    (right_unitor (obj X)).hom
+  = (𝟙 (obj X) ⊗ ε) ≫ μ X (tensor_unit C) ≫ map (right_unitor X).hom
+  . obviously)
+
+restate_axiom lax_monoidal_functor.μ_natural'
+attribute [simp] lax_monoidal_functor.μ_natural
+restate_axiom lax_monoidal_functor.left_unitality'
+attribute [simp] lax_monoidal_functor.left_unitality
+restate_axiom lax_monoidal_functor.right_unitality'
+attribute [simp] lax_monoidal_functor.right_unitality
+restate_axiom lax_monoidal_functor.associativity'
+attribute [simp] lax_monoidal_functor.associativity
+
+-- When `rewrite_search` lands, add @[search] attributes to
+-- lax_monoidal_functor.μ_natural lax_monoidal_functor.left_unitality
+-- lax_monoidal_functor.right_unitality lax_monoidal_functor.associativity
+
+structure monoidal_functor
+extends lax_monoidal_functor.{v₁ v₂} C D :=
+(ε_is_iso            : is_iso ε . obviously)
+(μ_is_iso            : Π X Y : C, is_iso (μ X Y) . obviously)
+
+attribute [instance] monoidal_functor.ε_is_iso monoidal_functor.μ_is_iso
+
+variables {C D}
+
+def monoidal_functor.ε_iso (F : monoidal_functor.{v₁ v₂} C D) :
+  tensor_unit D ≅ F.obj (tensor_unit C) :=
+as_iso F.ε
+def monoidal_functor.μ_iso (F : monoidal_functor.{v₁ v₂} C D) (X Y : C) :
+  (F.obj X) ⊗ (F.obj Y) ≅ F.obj (X ⊗ Y) :=
+as_iso (F.μ X Y)
+
+end
+
+open monoidal_category
+
+namespace monoidal_functor
+
+section
+-- In order to express the tensorator as a natural isomorphism,
+-- we need to be in at least `Type 0`, so we have products.
+variables {C : Type u₁} [𝒞 : monoidal_category.{v₁+1} C]
+variables {D : Type u₂} [𝒟 : monoidal_category.{v₂+1} D]
+include 𝒞 𝒟
+
+def μ_nat_iso (F : monoidal_functor.{v₁+1 v₂+1} C D) :
+  (functor.prod F.to_functor F.to_functor) ⋙ (tensor D) ≅ (tensor C) ⋙ F.to_functor :=
+nat_iso.of_components
+  (by intros; dsimp; apply F.μ_iso)
```

Comment:
Curious whether `dsimp` is really needed here (and other similar places); normally `apply` should be able to see through defeqs, though I guess maybe it doesn't always.

Reply:
Removed.